### ISSUE TITLE
Can no longer print, error on line 352 TypeError: string indices must be integers

### ISSUE THREAD

**Issue by fredrikchabot** [simoncadman/CUPS-Cloud-Print#96]
From /var/log/cups/error_log

Sending document to Cloud Print
Traceback (most recent call last):
File "/usr/lib/cups/backend/gcp", line 141, in <module>
elif printer.submitJob('pdf', printFile, filedata, jobTitle, cupsprintername, printOptions):
File "/usr/share/cloudprint-cups/printer.py", line 557, in submitJob
('capabilities', json.dumps(self._getCapabilities(cupsprintername, options)))
File "/usr/share/cloudprint-cups/printer.py", line 496, in _getCapabilities
return self._getCapabilitiesDict(attrArray, self['capabilities'], overridecapabilities)
File "/usr/share/cloudprint-cups/printer.py", line 445, in _getCapabilitiesDict
if hashname == Printer._getInternalName(capability, 'capability'):
File "/usr/share/cloudprint-cups/printer.py", line 352, in _getInternalName
if details['name'] == itemName:
TypeError: string indices must be integers
Backend returned status 1 (failed)
Printer stopped due to backend errors; please consult the error_log file for details.
End of messages
printer-state=5(stopped)
printer-state-message="/usr/lib/cups/backend/gcp failed"

I get the same error on two different servers, each with a different google account and different printers.

**Comment by ebarthelemyvxt** [comment]
Hello

We have exactly the same problem starting today 

[17/Mar/2016:15:47:28 +0100] [Job 57321] Sending document to Cloud Print
D [17/Mar/2016:15:47:28 +0100] [Job 57321] Set job-printer-state-message to "Sending document to Cloud Print", current level=INFO
D [17/Mar/2016:15:47:28 +0100] [Job 57321] Traceback (most recent call last):
D [17/Mar/2016:15:47:28 +0100] [Job 57321] File "/usr/lib/cups/backend/gcp", line 141, in <module>
D [17/Mar/2016:15:47:28 +0100] [Job 57321] elif printer.submitJob('pdf', printFile, filedata, jobTitle, cupsprintername, printOptions):
D [17/Mar/2016:15:47:28 +0100] [Job 57321] File "/usr/share/cloudprint-cups/printer.py", line 557, in submitJob
D [17/Mar/2016:15:47:28 +0100] [Job 57321] ('capabilities', json.dumps(self._getCapabilities(cupsprintername, options)))
D [17/Mar/2016:15:47:28 +0100] [Job 57321] File "/usr/share/cloudprint-cups/printer.py", line 496, in _getCapabilities
D [17/Mar/2016:15:47:28 +0100] [Job 57321] return self._getCapabilitiesDict(attrArray, self['capabilities'], overridecapabilities)
D [17/Mar/2016:15:47:28 +0100] [Job 57321] File "/usr/share/cloudprint-cups/printer.py", line 445, in _getCapabilitiesDict
D [17/Mar/2016:15:47:28 +0100] [Job 57321] if hashname == Printer._getInternalName(capability, 'capability'):
D [17/Mar/2016:15:47:28 +0100] [Job 57321] File "/usr/share/cloudprint-cups/printer.py", line 352, in _getInternalName
D [17/Mar/2016:15:47:28 +0100] [Job 57321] if details['name'] == itemName:
D [17/Mar/2016:15:47:28 +0100] [Job 57321] TypeError: string indices must be integers
D [17/Mar/2016:15:47:28 +0100] [Job 57321] Backend returned status 1 (failed)
D [17/Mar/2016:15:47:28 +0100] [Job 57321] End of messages
D [17/Mar/2016:15:47:28 +0100] [Job 57321] printer-state=3(idle)
D [17/Mar/2016:15:47:28 +0100] [Job 57321] printer-state-message="/usr/lib/cups/backend/gcp failed"
D [17/Mar/2016:15:47:28 +0100] [Job 57321] printer-state-reasons=none

**Comment by IBabitsch** [comment]
Yesterday service stoped and after reconfiguration same commented problem:
Can't get PPD file due following error (seems GCP changed service)

Traceback (most recent call last):
  File "/usr/share/cloudprint-cups/dynamicppd.py", line 107, in <module>
    doCat()
  File "/usr/share/cloudprint-cups/dynamicppd.py", line 64, in doCat
    print printer.generatePPD()
  File "/usr/share/cloudprint-cups/printer.py", line 280, in generatePPD
    self._getInternalName(capability, 'capability', None, addedCapabilities)
  File "/usr/share/cloudprint-cups/printer.py", line 352, in _getInternalName
    if details['name'] == itemName:
TypeError: string indices must be integers
Result of cupstestppd was 3
(stdin): FAIL
      **FAIL**  Unable to open PPD file - Illegal control character on line 1.
                REF: Page 15, section 3.1.

cupstestppd errored:

**Comment by Taitava** [comment]
I can confirm this, it stopped working yesterday. I have the same error in the log as @fredrikchabot .

**Comment by tavinus** [comment]
Yes, it is dead here too. Can't print anymore.
If I try to add a new printer I also get an error:
`Error adding: Save_to_Google_Drive (1280, 'server-error-internal-error')`

**Comment by Taitava** [comment]
@tavinus And if I try to add the same printer using USB connection, I also get the same error that you got when adding the Save to Google Drive printer. No matter what type of printer I'm trying to add, it fails. I didn't mention this at first because I was not sure if it was due to the actions I made after noticing that the printing does not work: I removed cupscloudprint, then reinstalled it, removed printers and tried to add them again with the cupscloudprint's config utility (didn't succeed), then removed and reinstalled cups. Perhaps made something else too, I can't remember everything now. Then I found this issue in Github and understood that it's not specific to my system only.

**Comment by simoncadman** [comment]
Hi All,

Sorry for the delay in responding, I've been dealing with a lot of personal & family issues lately which has distracted me from my projects. I'll see if I can find a fix for this - I'm also able to recreate this so it should be something I can fix.

It looks like something on Google's side has changed with how it returns capabilities ( format seems to have changed so we get an error when trying to process them ), adding this line to the end of __init__ inside the Printer class in printer.py then running upgrade.py at least makes basic printing work again:

self._fields['capabilities'] = {}

I'll see if I can update the code to work with the new capabilities format.

Thanks

**Comment by virtualdll** [comment]
confirmed the fix just run these two lines:

```
sed -i '173i        self._fields['capabilities'] = {}' /usr/share/cloudprint-cups/printer.py
/usr/share/cloudprint-cups/upgrade.py
```

if you have trouble still try to delete the printers and try again. 
You can remove account and printers by doing.
`/usr/share/cloudprint-cups/deleteaccount.py`

add them again by running 
`/usr/share/cloudprint-cups/setupcloudprint.py`

**Comment by tavinus** [comment]
Hi all. Thanks for the replies.

I can confirm that adding the provided line made it work again. I had to reinstall all printers though. Also, the sed command from virtualdll didn't work for me because it discarded the blank spaces (and then I got an identation error, which I fixed by just adding the white spaces needed). Only problem I noticed is that none of my printers provide the option for double-sided (duplex) anymore.

**Comment by jaymtee** [comment]
Adding the line to printer.py worked here.  Looking forward to the complete solution (and willing to test..)

**Comment by virtualdll** [comment]
Yeah sorry should have tested it a bit. The follow is working for me to apply the fix. 

sed -i "173iself._fields['capabilities'] = {}" /usr/share/cloudprint-cups/printer.py

/usr/share/cloudprint-cups/upgrade.py

**Comment by JuanAlberto99** [comment]
This workaround not work for me, still get the same error

**Comment by emnik** [comment]
same here, it doesn't work. I will post again if I'll find a different error in the logs as now I am not at pc with the CUPS-cloud-print....

**Comment by virtualdll** [comment]
What are the full steps you both did. 

Some are having to
Edit the file. (Sed works for some and some have to edit it manually)
Run the upgrade.py
Delete the printers
Add the printers
Restart cups

**Comment by sulli86** [comment]
**when trying to run **
/usr/share/cloudprint-cups/upgrade.py

**get the following error:**
Traceback (most recent call last):
  File "/usr/share/cloudprint-cups/upgrade.py", line 38, in <module>
    from printermanager import PrinterManager
  File "/usr/share/cloudprint-cups/printermanager.py", line 22, in <module>
    from printer import Printer
  File "/usr/share/cloudprint-cups/printer.py", line 175
    def getAccount(self):

**Comment by emnik** [comment]
I have edited printer.py and added the line. deleted the account, readded the printers and then:
root@tutors-office:/home/tutor/Apps/CUPS-Cloud-Print-master# ./upgrade.py 
Fetching list of available ppds...
List retrieved successfully
Updating GCP-__Google_Drive
lpadmin: Unable to copy PPD file.

root@tutors-office:/home/tutor/Apps/CUPS-Cloud-Print-master# cat /var/log/cups/error_log
E [29/Mar/2016:17:10:11 +0300] copy_model: empty PPD file
E [29/Mar/2016:17:10:11 +0300] [Client 20] Returning IPP server-error-internal-error for CUPS-Add-Modify-Printer (ipp://localhost:631/printers/GCP-__Google_Drive) from localhost

root@tutors-office:/home/tutor/Apps/CUPS-Cloud-Print-master# ./setupcloudprint.py 
You currently have these accounts configured: 
nikiforakis.m@gmail.com
Error adding: GCP-EPSON-Epson-Stylus-SX525WD
(1280, u'server-error-internal-error')
Error adding: GCP-EPSON-Epson-Stylus-SX525WD 
Error adding: GCP-Panasonic_DP-8016P_Ver_2
(1280, u'server-error-internal-error')
Error adding: GCP-Panasonic_DP-8016P_Ver_2 
Added 2 new printers to CUPS
it says added but they are not,...

**Comment by sulli86** [comment]
I resolved my issues,

in the printer.py file I found that the line i added was not indented with the rest of the coding, once i spaced it in it kicked into life and resolved the issue, thank goodness worked 8hours to find a fix for this.... customer is going to hate me

**Comment by tavinus** [comment]
Well, that is exactly what I mentioned in my post (indentation error after using the sed command provided). I also mentioned that all I had to do was to add the white spaces needed. That is normal for python and anyone with problems should check the modified source for an identation error. 

Have been using without problems ever since I changed that. I would love to have a complete solution that allows me to print double-sided again though (saves so much paper). I will keep monitoring here for news on that.

Cheers!

**Comment by virtualdll** [comment]
Yeah sorry about the sed command the one I posted originally worked for me with no issues. 

Thanks,
David Lamb

**Comment by simoncadman** [comment]
Believe double siding printing etc should work again with the changed in the capabilities-fix branch, doing some more testing but it seems to work so far with my printers.

Thanks

**Comment by tavinus** [comment]
Hi again!
I was taking a look at the capabilities-fix branch and wondering about it being 23 commits ahead and 94 commits behind master. Is it safe for me to install it over my current version? Why the 94 commits behind?

I would gladly test it, as we miss the double-sided printing a lot.

**Comment by simoncadman** [comment]
Hi All,

@tavinus It should be safe to install over your old version, it's just 94 commits behind master, as the master branch isn't really ready for release so I added the changes from the previous release version.

I've released a new version of CUPS Cloud Print ( 20160502 ), should all be working again with this release. I'd recommend installing the OS packages from https://ccp.niftiestsoftware.com rather than installing from source if possible though.

Thanks

**Comment by tavinus** [comment]
Hi! Thanks for the reply. I have been using the capabilities-fix branch for a week or two. I have compiled it from github. Had no problems to install or manage printers but could not print double-sided (using an HP that connects through HP ePrint). I will test with other printers (we have a few there), but so far no luck for me with that. 

Honestly, I could not test it very well and I am not frequently near those printers.

So, I guess I should reinstall everything again (with the new release, using yum)?
I had the old one uninstalled before compiling the new one. The main thing is that the default folder is different (/usr/share VS /usr/local/share). But had no problem to compile/install.

**Comment by simoncadman** [comment]
Hi,

I'd suggest installing using yum and delete or rename the /usr/local/share/cloudprint-cups/ directory, to ensure it's no longer in use.

I'll investigate further if there are still issues.

Thanks

**Comment by colorprint** [comment]
With new release installed I have another error when try to print anything:

"Print job pdf failed with 'ascii' codec can't encode characters in position 53-54: ordinal not in range(128)"

**Comment by tavinus** [comment]
Hi.. Sorry for posting on a closed topic, but this belongs here.

I spend some good hours testing many printers. After tweaking all the printer's configs I was able to get all of them working well, including double-sided printing. 

I realised that some printers would not even work because their tray was set to receive only A4-common paper and then GCP would send some files as Letter and the printer would just halt with an error, asking to change the tray/paper to Letter. That was also affecting double-sided printing. Setting stuff to "any-paper", "any-type", etc did the trick.

I am still using the compiled version. I am leaving the country tomorrow, so I will probably leave this for another 10 days or so. Seems to be quite stable anyways.

Thanks for the great work!
